test/e2e/specs/editor/various/block-bindings.spec.js

```diff
@@ -0,0 +1,856 @@
+									},
+								},
+							},
+						},
+					],
+				},
+			},
+			images: {
+				urlOnly: {},
+				altOnly: {},
+				titleOnly: {},
+				multipleAttrs: {},
+			},
+		},
+		placeholderSrc: '',
+	};
+	test.beforeAll( async ( { requestUtils } ) => {
+		await requestUtils.activateTheme( 'emptytheme' );
+		await requestUtils.activatePlugin( 'gutenberg-test-block-bindings' );
+		await requestUtils.deleteAllMedia();
+		const placeholderMedia = await requestUtils.uploadMedia(
+			path.join( './test/e2e/assets', '10x10_e2e_test_image_z9T8jK.png' )
+		);
+		variables.placeholderSrc = placeholderMedia.source_url;
+		// Init image blocks.
+		variables.blocks.images.urlOnly = {
+			name: 'core/image',
+			attributes: {
+				url: variables.placeholderSrc,
+				alt: 'default alt value',
+				title: 'default title value',
+				metadata: {
+					bindings: {
+						url: {
+							source: 'core/post-meta',
+							args: { key: 'url_custom_field' },
+						},
+					},
+				},
+			},
+		};
+		variables.blocks.images.altOnly = {
+			name: 'core/image',
+			attributes: {
+				url: variables.placeholderSrc,
+				alt: 'default alt value',
+				title: 'default title value',
+				metadata: {
+					bindings: {
+						alt: {
+							source: 'core/post-meta',
+							args: { key: 'text_custom_field' },
+						},
+					},
+				},
+			},
+		};
+		variables.blocks.images.titleOnly = {
+			name: 'core/image',
+			attributes: {
+				url: variables.placeholderSrc,
+				alt: 'default alt value',
+				title: 'default title value',
+				metadata: {
+					bindings: {
+						title: {
+							source: 'core/post-meta',
+							args: { key: 'text_custom_field' },
+						},
+					},
+				},
+			},
+		};
+		variables.blocks.images.multipleAttrs = {
+			name: 'core/image',
+			attributes: {
+				url: variables.placeholderSrc,
+				alt: 'default alt value',
+				title: 'default title value',
+				metadata: {
+					bindings: {
+						url: {
+							source: 'core/post-meta',
+							args: { key: 'url_custom_field' },
+						},
+						alt: {
+							source: 'core/post-meta',
+							args: { key: 'text_custom_field' },
+						},
+					},
+				},
+			},
+		};
+	} );
+
+	test.afterEach( async ( { requestUtils } ) => {
+		await requestUtils.deleteAllPosts();
+	} );
+
+	test.afterAll( async ( { requestUtils } ) => {
+		await requestUtils.deleteAllMedia();
+		await requestUtils.activateTheme( 'twentytwentyone' );
+		await requestUtils.deactivatePlugin( 'gutenberg-test-block-bindings' );
+	} );
+
+	test.describe( 'Template context', () => {
+		test.beforeEach( async ( { admin, editor } ) => {
+			await admin.visitSiteEditor( {
+				postId: 'emptytheme//index',
+				postType: 'wp_template',
+			} );
+			await editor.canvas.locator( 'body' ).click();
+			await editor.openDocumentSettingsSidebar();
+		} );
+
+		test.describe( 'Paragraph', () => {
+			test( 'Should show the value of the custom field', async ( {
+				editor,
+			} ) => {
+				await editor.insertBlock( variables.blocks.paragraph );
```

Comment:
Nit: IMHO, I think these "variables" are less readable because I have to scroll up and down this file to know exactly what it is. I would probably just inline this, or make a function if it's used several times. These are tests, so I wouldn't worry too much about repetitions.

Reply:
I was more worried about the maintenance of the tests. Anyway, I addressed that in this commit as part of a follow-up pull request: [link](https://github.com/WordPress/gutenberg/commit/9842c156e59f84df5b5cfd7801e120a8903e23ca).